### DISCUSSION TITLE
OpenTelemetry Integration

### DISCUSSION THREAD

**Discussion by cwinter158** [capnproto/capnproto#2165] [Q&A] [answered]
I'm looking to trace some capnproto calls using OpenTelemetry; in order to trace calls that cross multiple servers (client A -> server B -> server C) and join everything up, I need to propogate some metadata (a trace id and span id) alongside the RPC calls. Looking through the documentation/headers, I'm not seeing anything jump out as a way to attach some additional metadata, but maybe I missed something?

Alternatively, do you have any suggestions for how to do this, other than just attaching the metadata as part of the RPC request contents (which I'm not necessarily opposed to, just wonder if there is a better way)?

Thanks in advance!

**Comment by kentonv** [comment, marked as answer]
At present there's no built-in way to send metadata like this, you have to make it part of the RPC params. A technique I have used, though, is to have two-layer RPC, like:

```
interface Traced(T) {
  # Wrapper around a service to add tracing support.

  startCall @0 (traceId :Text) -> (service :T);
  # Provides trace ID for a call. After calling startCall(), the caller should used the
  # returned `service` to make exactly one call.
}
```

This way, you don't actually have to add `traceId` as a parameter to every single RPC method of your service -- you use `Traced(MyService)` instead.

That said, I think it would make sense to add some support for trace IDs directly to Cap'n Proto at some point.

**Reply by cwinter158** [reply]
Ah ok, got it. That's an interesting technique, I'll have to give it a try. Thanks!
